compiler/circle-mpqsolver/src/ErrorApproximator.cpp

```diff
@@ -0,0 +1,341 @@
+                                   int32_t &channel_dim_index)
+{
+  loco::TensorShape dimension;
+  dimension.rank(4);
+
+  if (!get_channel_dim_index(node, dimension, channel_dim_index))
+  {
+    throw std::runtime_error("Failed to find channel index in " + node->name());
+  }
+  auto size = dimension.dim(channel_dim_index).value();
+
+  std::vector<bool> has_min_max_value(size, false);
+  values.resize(size);
+
+  auto cal_minmax = [&](uint32_t *indices, loco::TensorShape &dimension, int channel_dim_index) {
+    int channel_idx = indices[channel_dim_index];
+    auto data = node->at<loco::DataType::FLOAT32>(cal_offset(dimension, indices));
+    values[channel_idx].push_back(data);
+  };
+
+  iterate_per_channel(node, channel_dim_index, cal_minmax);
+}
+} // namespace
+
+/**
+ * How Approximate works?
+ *
+ * Currently it works just for convolution layers, but may be generalized for other types as well.
+ * See discussion at https://github.com/Samsung/ONE/pull/10170#discussion_r1042246598
+ * Convolution is just a matrix multiplication.
+ * While quantizing we intoduce quantization error into convolution operand (activations) as well as
+ * into convolution weights. A_q * W_q = (A + q_err(A)) * (W + q_err(W)) = A * W + A * q_err(W) + W
+ * * q_err(A), q_err - quantization error, W - weight matrix, A - activations from previous layer,
+ * so quantization error of matrix multiplication is A * q_err(W) + W * q_err(A).
+ * Estimating its largest value we get
+ * ~ number_of_additions * (A_max * (W_max - W_min) / 255 + W_max * (A_max - A_min) / 255)
+ * The following code tries to get total error for quantizing convolution node into Q8.
+ * It's just an heuristic (Metric sensitivity depends highly on derivatives as well).
+ */
+float ErrorApproximator::Approximate(CircleNode *node)
+{
+  auto circle_node = loco::must_cast<luci::CircleNode *>(node);
+  auto name = circle_node->name();
+  CircleConst *weights = nullptr;
+  uint32_t volume = 0;
+  std::vector<uint32_t> cir_shape;
+  if (getShape(circle_node, cir_shape) && cir_shape.size() == 4)
+  {
+    volume = cir_shape[1] * cir_shape[2] * cir_shape[0];
+  }
+
+  if (auto convWeights = dynamic_cast<luci::CircleConv2D *>(circle_node))
+  {
+    weights = loco::must_cast<luci::CircleConst *>(convWeights->filter());
+  }
+  else if (auto convWeights = dynamic_cast<luci::CircleDepthwiseConv2D *>(circle_node))
+  {
+    weights = loco::must_cast<luci::CircleConst *>(convWeights->filter());
+  }
+  else if (auto convWeights = dynamic_cast<luci::CircleTransposeConv *>(circle_node))
+  {
+    weights = loco::must_cast<luci::CircleConst *>(convWeights->filter());
+  }
+
+  if (weights != nullptr)
+  {
+    float ci_min = 0.f;
+    float ci_max = 0.f;
+
+    int n_arity = circle_node->arity();
+    for (int i = 0; i < n_arity; i++)
+    {
+      auto parent_node = loco::must_cast<luci::CircleNode *>(circle_node->arg(i));
```

Comment:
You can use `loco::preds` method.
```suggestion
    auto preds = loco::preds(n);
    for (const auto &pred : preds)
    {
      auto parent_node = loco::must_cast<const luci::CircleNode *>(pred);
```

Reply:
Ahhh... Ok. Thank you!